File: trunk/detection/executor/cpp/streaming/StandardInWatcher.cpp
/******************************************************************************
 * NOTICE                                                                     *
 *                                                                            *
 * This software (or technical data) was produced for the U.S. Government     *
 * under contract, and is subject to the Rights in Data-General Clause        *
 * 52.227-14, Alt. IV (DEC 2007).                                             *
 *                                                                            *
 * Copyright 2017 The MITRE Corporation. All Rights Reserved.                 *
 ******************************************************************************/

/******************************************************************************
 * Copyright 2017 The MITRE Corporation                                       *
 *                                                                            *
 * Licensed under the Apache License, Version 2.0 (the "License");            *
 * you may not use this file except in compliance with the License.           *
 * You may obtain a copy of the License at                                    *
 *                                                                            *
 *    http://www.apache.org/licenses/LICENSE-2.0                              *
 *                                                                            *
 * Unless required by applicable law or agreed to in writing, software        *
 * distributed under the License is distributed on an "AS IS" BASIS,          *
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   *
 * See the License for the specific language governing permissions and        *
 * limitations under the License.                                             *
 ******************************************************************************/


#include <functional>
#include <iostream>
#include <thread>

#include "StandardInWatcher.h"
#include "ExecutorErrors.h"

namespace MPF { namespace COMPONENT {
    std::atomic_bool StandardInWatcher::quit_received_(false);
    std::string StandardInWatcher::error_message_;

    StandardInWatcher* StandardInWatcher::instance_(nullptr);


    StandardInWatcher::StandardInWatcher() {
        std::thread watcher_thread(&Watch);
        // detach so that when there is an error elsewhere, the process can exit without waiting for the quit command.
        watcher_thread.detach();
    }


    bool StandardInWatcher::QuitReceived() const {
        if (quit_received_ && !error_message_.empty()) {
            throw FatalError(ExitCode::UNABLE_TO_READ_FROM_STANDARD_IN, error_message_);
        }
        return quit_received_;
    }


    void StandardInWatcher::Watch() {
        try {
            std::string line;
            while (std::getline(std::cin, line)) {
                if (line == "quit") {
                    quit_received_ = true;
                    return;
                }
                // TODO: Check for "pause" and "resume" when adding support for multistage pipelines.
                std::cerr << "Ignoring unexpected input from standard in: " << line << std::endl;
            }

            if (std::cin.eof()) {
                SetError("Standard in was closed before quit was received.");
            }
            else {
                SetError("An error occurred while trying to read from standard in.");
            }
        }
        catch (const std::exception &ex) {
            SetError(std::string("An error occurred while trying to read from standard in: ") + ex.what());
        }
        catch (...) {
            SetError("An error occurred while trying to read from standard in.");
        }
    }


    StandardInWatcher* StandardInWatcher::GetInstance() {
        if (instance_ == nullptr) {
            instance_ = new StandardInWatcher;
        }
        return instance_;
    }

    void StandardInWatcher::SetError(std::string &&error_message) {
        // error_message_ must be set before is_time_to_quit_ to ensure proper synchronization.
        // error_message_ is not atomic because there is no atomic string.
        // Since is_time_to_quit_ is atomic, reading or writing to is_time_to_quit_ will cause a synchronization event.
        error_message_ = std::move(error_message);
        quit_received_ = true;
    }
}}

File: trunk/markup/src/main/java/org/mitre/mpf/markup/StandardInWatcher.java
/******************************************************************************
 * NOTICE                                                                     *
 *                                                                            *
 * This software (or technical data) was produced for the U.S. Government     *
 * under contract, and is subject to the Rights in Data-General Clause        *
 * 52.227-14, Alt. IV (DEC 2007).                                             *
 *                                                                            *
 * Copyright 2023 The MITRE Corporation. All Rights Reserved.                 *
 ******************************************************************************/

/******************************************************************************
 * Copyright 2023 The MITRE Corporation                                       *
 *                                                                            *
 * Licensed under the Apache License, Version 2.0 (the "License");            *
 * you may not use this file except in compliance with the License.           *
 * You may obtain a copy of the License at                                    *
 *                                                                            *
 *    http://www.apache.org/licenses/LICENSE-2.0                              *
 *                                                                            *
 * Unless required by applicable law or agreed to in writing, software        *
 * distributed under the License is distributed on an "AS IS" BASIS,          *
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   *
 * See the License for the specific language governing permissions and        *
 * limitations under the License.                                             *
 ******************************************************************************/

package org.mitre.mpf.markup;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

import javax.jms.Connection;
import javax.jms.JMSException;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class StandardInWatcher implements Runnable {

    private static final Logger LOG = LoggerFactory.getLogger(StandardInWatcher.class);

    private volatile boolean _quitReceived = false;

    private final Connection _amqConnection;

    private final Thread _messageProcessingThread;


    public static StandardInWatcher start(
            Connection amqConnection, Thread messageProcessingThread) {
        var watcher = new StandardInWatcher(amqConnection, messageProcessingThread);
        var thread = new Thread(watcher, "StandardInWatcher");
        thread.setDaemon(true);
        thread.start();
        return watcher;
    }


    private StandardInWatcher(
            Connection amqConnection, Thread messageProcessingThread) {
        _amqConnection = amqConnection;
        _messageProcessingThread = messageProcessingThread;
    }

    public boolean quitReceived() {
        return _quitReceived;
    }

    @Override
    public void run() {
        try (var reader = new BufferedReader(new InputStreamReader(System.in))) {
            String line;
            while ((line = reader.readLine()) != null) {
                LOG.info("Received input on stdin: \"{}\"", line);
                if (line.startsWith("q")) {
                    _quitReceived = true;
                    shutdown();
                    return;
                }
            }
            LOG.info("Standard in was closed. Must use a signal to exit.");
        }
        catch (IOException e) {
            LOG.error(e.getMessage(), e);
        }
    }

    private void shutdown() {
        try {
            LOG.info("Closing ActiveMQ connection...");
            // Closing the connection should cause the message processing thread to exit.
            // If the message processing thread is waiting for a message, the receive method will
            // return null. If the message processing thread is currently processing a message, an
            // exception will be thrown the next time it calls a JMS method.
            _amqConnection.close();
            LOG.info("ActiveMQ connection closed.");
        }
        catch (JMSException e) {
            LOG.error("An error occurred while trying to close ActiveMQ connection: " + e, e);
        }

        try {
            _messageProcessingThread.join(1_000);
            if (_messageProcessingThread.isAlive()) {
                LOG.info("Message processing thread did not exit when connection closed, attempting interrupt.");
                _messageProcessingThread.interrupt();
            }
        }
        catch (InterruptedException ignored) {
            // Already shutting down
            Thread.currentThread().interrupt();
        }
    }
}
